src/asset-management/AaveManager.sol

```diff
@@ -96,39 +98,72 @@ contract AaveManager is IAssetManager, Ownable, ReentrancyGuard
         }
         if (aAmount1Change > 0) {
-            _doInvest(aPair, lToken1, uint256(aAmount1Change));
+            lInvestFail1 = _doInvest(aPair, lToken1, lToken1AToken, uint256(aAmount1Change));
         }
-    }
 
-    function _doDivest(IAssetManagedPair aPair, IERC20 aToken, uint256 aAmount) private {
-        uint256 lShares = _updateShares(aPair, address(aToken), aAmount, false);
-        pool.withdraw(address(aToken), aAmount, address(this));
-        emit FundsDivested(aPair, address(aToken), lShares);
+        // invest failed, put the money back in the pair
+        if (lInvestFail0 || lInvestFail1) {
+            // update the pair about the updated amount
+            aPair.adjustManagement(-int256(lInvestFail0 ? aAmount0Change : int256(0)), -int256(lInvestFail1 ? aAmount1Change : int256(0)));
+        }
+    }
 
-        aToken.approve(address(aPair), aAmount);
+    function _doDivest(IAssetManagedPair aPair, IERC20 aToken, address aAaveToken, uint256 aAmount) private returns (uint256 rActualWithdrawn) {
+        uint256 lShares = _updateShares(aPair, address(aToken), aAaveToken, aAmount, false);
+        try pool.withdraw(address(aToken), aAmount, address(this)) returns (uint256 rAmountWithdrawn) {
+            emit FundsDivested(aPair, aToken, lShares);
+            aToken.approve(address(aPair), aAmount);
+            rActualWithdrawn = rAmountWithdrawn;
+            // might take this out in the future
+            assert(rActualWithdrawn == aAmount);
+        }
+        catch {
+            rActualWithdrawn = 0;
+            // undo the shares change
+            shares[aPair][address(aToken)] += lShares;
+            totalShares[aAaveToken] += lShares;
+        }
     }
 
-    function _doInvest(IAssetManagedPair aPair, IERC20 aToken, uint256 aAmount) private {
+    function _doInvest(IAssetManagedPair aPair, IERC20 aToken, address aAaveToken, uint256 aAmount) private returns (bool rFail) {
         require(aToken.balanceOf(address(this)) == aAmount, "AM: TOKEN_AMOUNT_MISMATCH");
-
-        uint256 lShares = _updateShares(aPair, address(aToken), aAmount, true);
+        uint256 lShares = _updateShares(aPair, address(aToken), aAaveToken, aAmount, true);
         aToken.approve(address(pool), aAmount);
-        pool.supply(address(aToken), aAmount, address(this), 0);
-        emit FundsInvested(aPair, address(aToken), lShares);
+
+        try pool.supply(address(aToken), aAmount, address(this), 0) {
+            emit FundsInvested(aPair, aToken, lShares);
+            rFail = false;
+        }
+        // if the supply fails for whatever reason
+        // pool could be frozen, paused, reached its supply limit etc
+        catch {
+            // approve the token for returning to the pair
+            aToken.approve(address(aPair), aAmount);
+            rFail = true;
+            // undo shares change
+            shares[aPair][address(aToken)] -= lShares;
+            totalShares[aAaveToken] -= lShares;
+        }
```

Comment:
I think all these changes are in the wrong PR? They are not directly related to `returnAsset` right? Can we remove them from this PR?